guide/src/python-from-rust.md

```diff
@@ -1,511 +1,45 @@
 
-        house.call_method0("__enter__").unwrap();
+PyO3's types that are bound to the `'py` lifetime, for example `Bound<'py, T>`, all contain a `Python<'py>` token. This means they have full access to the Python interpreter and offer a complete API for interacting with Python objects.
 
-        let result = py.eval_bound("undefined_variable + 1", None, None);
+Consult [PyO3's API documentation][obtaining-py] to learn how to acquire one of these tokens.
 
-        // If the eval threw an exception we'll pass it through to the context manager.
-        // Otherwise, __exit__  is called with empty arguments (Python "None").
-        match result {
-            Ok(_) => {
-                let none = py.None();
-                house
-                    .call_method1("__exit__", (&none, &none, &none))
-                    .unwrap();
-            }
-            Err(e) => {
-                house
-                    .call_method1("__exit__", (e.get_type_bound(py), e.value_bound(py), e.traceback_bound(py)))
-                    .unwrap();
-            }
-        }
-    })
-}
-```
+### The Global Interpreter Lock
 
-## Handling system signals/interrupts (Ctrl-C)
+Concurrent programming in Python is aided by the Global Interpreter Lock (GIL), which ensures that only one Python thread can use the Python interpreter and its API at the same time. This allows it to be used to synchronize code. See the [`pyo3::sync`] module for synchronization tools PyO3 offers that are based on the GIL's guarantees.
 
-The best way to handle system signals when running Rust code is to periodically call `Python::check_signals` to handle any signals captured by Python's signal handler. See also [the FAQ entry](./faq.md#ctrl-c-doesnt-do-anything-while-my-rust-code-is-executing).
+Non-Python operations (system calls and native Rust code) can unlock the GIL. See [the section on parallelism](parallelism.md) for how to do that using PyO3's API.
 
-Alternatively, set Python's `signal` module to take the default action for a signal:
+## Python's memory model
 
-```rust
-use pyo3::prelude::*;
+Python's memory model differs from Rust's memory model in two key ways:
+- There is no concept of ownership; all Python objects are reference counted
+- There is no concept of exclusive (`&mut`) references; any reference can mutate a Python object
 
-# fn main() -> PyResult<()> {
-Python::with_gil(|py| -> PyResult<()> {
-    let signal = py.import_bound("signal")?;
-    // Set SIGINT to have the default action
-    signal
-        .getattr("signal")?
-        .call1((signal.getattr("SIGINT")?, signal.getattr("SIG_DFL")?))?;
-    Ok(())
-})
-# }
-```
+PyO3's API reflects this by providing smart pointer types, `Py<T>`, `Bound<'py, T>`, and (the very rarely used) `Borrowed<'a, 'py, T>`. These smart pointers all use Python reference counting. See the [subchapter on types](./types.md) for more detail on these types.
```

Comment:
I wonder if it would make sense to refer to the analogy of `Py<T>` et al. with `Arc<Mutex<T>>`, but using the GIL instead of per-object mutexes.

Reply:
Good question. Given that Python 3.13 is coming around the corner the freethreaded build, I struggled in a few places to decide how much to go into detail about existing semantics vs comment on what's coming. Here, for example, objects will use "critical sections" which is probably much closer analogy to `Arc<Mutex<T>>` than the GIL currently is.

Would it be acceptable to punt on this analogy for now and improve further when I start updating for `nogil`? (Which I would like to start looking at next after 0.21 is "done" and hopefully sometime in April will get into full swing...)

Reply:
> Would it be acceptable to punt on this analogy for now and improve further when I start updating for nogil? (Which I would like to start looking at next after 0.21 is "done" and hopefully sometime in April will get into full swing...)

Certainly. As a simpler action to help people match this up with Rust native concepts, how about just making "smart pointer" a link into the book's chapter 15: <https://doc.rust-lang.org/book/ch15-00-smart-pointers.html>?

Reply:
Great idea 👍